src/assets/CurrencyRateController.ts

```diff
@@ -1,184 +1,216 @@
 
-/**
- * @type CurrencyRateConfig
- *
- * Currency rate controller configuration
- *
- * @property currentCurrency - Currently-active ISO 4217 currency code
- * @property interval - Polling interval used to fetch new currency rate
- * @property nativeCurrency - Symbol for the base asset used for conversion
- * @property includeUSDRate - Whether to include the usd rate in addition to the currentCurrency
- */
-export interface CurrencyRateConfig extends BaseConfig {
-  currentCurrency: string;
-  interval: number;
-  nativeCurrency: string;
-  includeUSDRate?: boolean;
-}
+import type { RestrictedControllerMessenger } from '../ControllerMessenger';
 
 /**
  * @type CurrencyRateState
  *
- * Currency rate controller state
- *
  * @property conversionDate - Timestamp of conversion rate expressed in ms since UNIX epoch
  * @property conversionRate - Conversion rate from current base asset to the current currency
  * @property currentCurrency - Currently-active ISO 4217 currency code
  * @property nativeCurrency - Symbol for the base asset used for conversion
+ * @property pendingCurrentCurrency - The currency being switched to
+ * @property pendingNativeCurrency - The base asset currency being switched to
  * @property usdConversionRate - Conversion rate from usd to the current currency
  */
-export interface CurrencyRateState extends BaseState {
+export type CurrencyRateState = {
   conversionDate: number;
   conversionRate: number;
   currentCurrency: string;
   nativeCurrency: string;
-  usdConversionRate?: number;
-}
+  pendingCurrentCurrency: string | null;
+  pendingNativeCurrency: string | null;
+  usdConversionRate: number | null;
+};
+
+const name = 'CurrencyRateController';
+
+export type CurrencyRateStateChange = {
+  type: `${typeof name}:stateChange`;
+  payload: [CurrencyRateState, Patch[]];
+};
+
+const metadata = {
+  conversionDate: { persist: true, anonymous: true },
+  conversionRate: { persist: true, anonymous: true },
+  currentCurrency: { persist: true, anonymous: true },
+  nativeCurrency: { persist: true, anonymous: true },
+  pendingCurrentCurrency: { persist: false, anonymous: true },
+  pendingNativeCurrency: { persist: false, anonymous: true },
+  usdConversionRate: { persist: true, anonymous: true },
+};
+
+const defaultState = {
+  conversionDate: 0,
+  conversionRate: 0,
+  currentCurrency: 'usd',
+  nativeCurrency: 'ETH',
+  pendingCurrentCurrency: null,
+  pendingNativeCurrency: null,
+  usdConversionRate: null,
+};
 
 /**
  * Controller that passively polls on a set interval for an exchange rate from the current base
  * asset to the current currency
  */
 export class CurrencyRateController extends BaseController<
-  CurrencyRateConfig,
+  typeof name,
   CurrencyRateState
 > {
-  /* Optional config to include conversion to usd in all price url fetches and on state */
-  includeUSDRate?: boolean;
+  private mutex = new Mutex();
 
-  private activeCurrency = '';
+  private handle?: NodeJS.Timer;
```

Comment:
`handle` is a confusing name for a timer. This is probably a good opportunity to change it. Maybe `pollingInterval`?

Reply:
Whoops; this used to be a `setTimeout` handler that would keep getting reset each iteration to point at the new handler. I updated it to use an interval but forgot to rename this. I'll change it to `interval`.

Reply:
Done in https://github.com/MetaMask/controllers/pull/372/commits/3965af46c05a638993ac899a371b6cee5c61963f